src/Tests/Microsoft.Diagnostics.Monitoring.ConfigurationSchema/SchemaGenerator.cs

```diff
@@ -57,64 +48,111 @@ public string GenerateSchema()
+        private static void AddDefaultConsoleLoggerOptionsSubSchema(GenerationContext context)
         {
             JsonSchema consoleLoggerOptionsSchema = new JsonSchema();
 
             JsonSchemaProperty formatterNameProperty = new JsonSchemaProperty();
             JsonSchemaProperty formatterOptionsProperty = new JsonSchemaProperty();
-            
-            JsonSchema formatterOptionsSchema = new JsonSchema();
-            formatterOptionsSchema.Reference = consoleFormatterOptions;
 
-            formatterOptionsProperty.Reference = formatterOptionsSchema;
+            formatterOptionsProperty.Reference = context.AddTypeIfNotExist<SimpleConsoleFormatterOptions>();
 
             formatterNameProperty.Type = JsonObjectType.Null;
             formatterNameProperty.Default = "Simple";
 
             consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterName), formatterNameProperty);
             consoleLoggerOptionsSchema.Properties.Add(nameof(ConsoleLoggerOptions.FormatterOptions), formatterOptionsProperty);
 
-            return consoleLoggerOptionsSchema;
+            context.Schema.Definitions[nameof(ConsoleLoggerOptions)].OneOf.Add(consoleLoggerOptionsSchema);
+        }
+
+        private static JsonSchemaProperty AddDiscriminatedSubSchema(
+            JsonSchema parentSchema,
+            string descriminatingPropertyName,
+            string descriminatingPropertyValue,
+            string descriminatedPropertyName,
+            JsonSchema subSchema = null)
+        {
+            if (null == subSchema)
+            {
+                subSchema = new JsonSchema();
+            }
+
+            JsonSchemaProperty descriminatingProperty = new JsonSchemaProperty();
+            descriminatingProperty.ExtensionData = new Dictionary<string, object>();
+            descriminatingProperty.ExtensionData.Add("const", descriminatingPropertyValue);
+
+            subSchema.Properties.Add(descriminatingPropertyName, descriminatingProperty);
+
+            JsonSchemaProperty descriminatedProperty = new JsonSchemaProperty();
+
+            subSchema.Properties.Add(descriminatedPropertyName, descriminatedProperty);
+
+            parentSchema.OneOf.Add(subSchema);
+
+            return descriminatedProperty;
+        }
+
+        private class GenerationContext
+        {
+            private readonly JsonSchemaGenerator _generator;
+            private readonly JsonSchemaResolver _resolver;
+            private readonly JsonSchemaGeneratorSettings _settings;
+
+            public GenerationContext(JsonSchema rootSchema)
+            {
+                Schema = rootSchema;
+
+                _settings = new JsonSchemaGeneratorSettings();
+                _settings.SerializerSettings = new JsonSerializerSettings();
+                _settings.SerializerSettings.Converters.Add(new StringEnumConverter());
+
+                _resolver = new JsonSchemaResolver(rootSchema, _settings);
```

Comment:
Supposedly JsonSchemaResolver is now considered obsolete; does this change anything?

Reply:
Where do you see that it is obsolete? There's no Obsolete attribute on it, no compiler warnings/errors, and I haven't seen that in documentation.

Reply:
Maybe I'm looking at the wrong thing? This is what I found when I searched: https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Schema_JsonSchemaResolver.htm (it says at the top `Note: This API is now obsolete`).

Reply:
> Maybe I'm looking at the wrong thing? This is what I found when I searched: https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Schema_JsonSchemaResolver.htm (it says at the top `Note: This API is now obsolete`).

The schema generation uses [NJsonSchema](https://github.com/RicoSuter/NJsonSchema), not Newtonsoft.Json. While NJsonSchema uses NewtonsoftJson as a dependency, this schema generation uses the NJsonSchema API.